<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 2: Analyzing Chipotle Data

_Author: Joseph Nelson (DC)_

---

For Project 2, you will complete a series of exercises exploring [order data from Chipotle](https://github.com/TheUpshot/chipotle), compliments of _The New York Times'_ "The Upshot."

For these exercises, you will conduct basic exploratory data analysis (Pandas not required) to understand the essentials of Chipotle's order data: how many orders are being made, the average price per order, how many different ingredients are used, etc. These allow you to practice business analysis skills while also becoming comfortable with Python.

---

## Basic Level

### Part 1: Read in the file with `csv.reader()` and store it in an object called `file_nested_list`.

Hint: This is a TSV (tab-separated value) file, and `csv.reader()` needs to be told [how to handle it](https://docs.python.org/2/library/csv.html).

In [59]:
import pandas as pd
import numpy as np

In [3]:
file_nested_list = pd.read_csv('data/chipotle.tsv', sep='\t')

In [5]:
file_nested_list.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

### Part 2: Separate `file_nested_list` into the `header` and the `data`.


In [12]:
header = file_nested_list.columns
header

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [13]:
data = file_nested_list.values
data

array([[1, 1, 'Chips and Fresh Tomato Salsa', nan, '$2.39 '],
       [1, 1, 'Izze', '[Clementine]', '$3.39 '],
       [1, 1, 'Nantucket Nectar', '[Apple]', '$3.39 '],
       ..., 
       [1834, 1, 'Chicken Salad Bowl',
        '[Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]',
        '$11.25 '],
       [1834, 1, 'Chicken Salad Bowl',
        '[Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]', '$8.75 '],
       [1834, 1, 'Chicken Salad Bowl',
        '[Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]',
        '$8.75 ']], dtype=object)

---

## Intermediate Level

### Part 3: Calculate the average price of an order.

Hint: Examine the data to see if the `quantity` column is relevant to this calculation.

Hint: Think carefully about the simplest way to do this!

In [14]:
# Make a copy of the file_nested_list that is shorter to deal with it going foward
ds = file_nested_list

In [17]:
# Check on the data tupes and see if order is numeric without missing values
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null object
dtypes: int64(2), object(3)
memory usage: 180.6+ KB


In [23]:
ds.isnull().any()

order_id              False
quantity              False
item_name             False
choice_description     True
item_price            False
dtype: bool

It seems price does not have any missing values but it a String because of the dollar sign. We need to remove the dollar sign and then cast it as float

In [32]:
ds['item_price'] = pd.to_numeric(ds['item_price'].str[1:])

In [33]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.6+ KB


The iem_price column has been sucessfully been transformed into a float and the mean per order can now be calculated.

In [91]:
print(ds.groupby('order_id').sum().item_price.mean().round(2))

18.81


From the table above the mean of the mean iem price is **$18.8*

### Part 4: Create a list (or set) named `unique_sodas` containing all of unique sodas and soft drinks that Chipotle sells.

Note: Just look for `'Canned Soda'` and `'Canned Soft Drink'`, and ignore other drinks like `'Izze'`.

In [72]:
ds.head(3)

order_id  quantity                     item_name choice_description  \
0         1         1  Chips and Fresh Tomato Salsa                NaN   
1         1         1                          Izze       [Clementine]   
2         1         1              Nantucket Nectar            [Apple]   

   item_price  
0       $2.39  
1       $3.39  
2       $3.39

In [100]:
unique_sodas = ds[(ds['item_name'] == 'Canned Soda') | (ds['item_name'] == 'Canned Soft Drink')]

order_id  quantity          item_name choice_description  item_price
18          9         2        Canned Soda           [Sprite]       $2.18
28         14         1        Canned Soda       [Dr. Pepper]       $1.09
51         23         2        Canned Soda     [Mountain Dew]       $2.18
53         24         1        Canned Soda           [Sprite]       $1.09
107        47         1        Canned Soda       [Dr. Pepper]       $1.09
117        51         1        Canned Soda  [Diet Dr. Pepper]       $1.09
126        55         1        Canned Soda        [Coca Cola]       $1.09
162        73         2        Canned Soda        [Diet Coke]       $2.18
171        76         2        Canned Soda  [Diet Dr. Pepper]       $2.18
179        81         1        Canned Soda        [Coca Cola]       $1.09
180        81         1        Canned Soda       [Dr. Pepper]       $1.09
195        87         1        Canned Soda        [Coca Cola]       $1.09
200        89         1        Canned Soda        [Diet Coke]       $1.09
228       100         1        Canned Soda     [Mountain Dew]       $1.09
246       108         1        Canned Soda     [Mountain Dew]       $1.09
247       108         1        Canned Soda       [Dr. Pepper]       $1.09
248       108         1        Canned Soda     [Mountain Dew]       $1.09
251       109         1        Canned Soda  [Diet Dr. Pepper]       $1.09
261       113         1        Canned Soda     [Mountain Dew]       $1.09
263       114         1  Canned Soft Drink             [Coke]       $1.25

In [122]:
print('Percentage of consumed softdrinks')
print('Canned Soda',round((unique_sodas.groupby('item_name').quantity.sum()[0]/ds.quantity.sum())*100,1),'%')
print('Canned Soft Drink',round((unique_sodas.groupby('item_name').quantity.sum()[1]/ds.quantity.sum())*100,1),'%')

Percentage of consumed softdrinks
Canned Soda 2.5 %
Canned Soft Drink 7.1 %


In [157]:
unique_sodas.choice_description.value_counts()

[Diet Coke]          110
[Coke]               102
[Sprite]              72
[Lemonade]            31
[Coca Cola]           26
[Nestea]              18
[Dr. Pepper]          18
[Mountain Dew]        15
[Diet Dr. Pepper]     13
Name: choice_description, dtype: int64

---

## Advanced Level


### Part 5: Calculate the average number of toppings per burrito.

Note: Let's ignore the `quantity` column to simplify this task.

Hint: Think carefully about the easiest way to count the number of toppings!


In [153]:
ds.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                                NaN       $2.39  
1                                       [Clementine]       $3.39  
2                                            [Apple]       $3.39  
3                                                NaN       $2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...      $16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...      $10.98  
6                                                NaN       $1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...      $11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...       $9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...       $9.25

In [209]:
# Get all the choices just for Burritos
burit_topp = ds[ds.item_name.str.contains('burrito', case=False)].choice_description

cnt = 0
for item in burit_topp:
    i = item.split(',')
    cnt += len(i)
print('Average Number Burritos per topping:', round(cnt / burit_topp.count(),2))

Average Number Burritos per topping: 5.4


### Part 6: Create a dictionary. Let the keys represent chip orders and the values represent the total number of orders.

Expected output: `{'Chips and Roasted Chili-Corn Salsa': 18, ... }`

Note: Please take the `quantity` column into account!

Optional: Learn how to use `.defaultdict()` to simplify your code.

In [210]:
ds.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description  item_price  
0                                                NaN       $2.39  
1                                       [Clementine]       $3.39  
2                                            [Apple]       $3.39  
3                                                NaN       $2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...      $16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...      $10.98  
6                                                NaN       $1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...      $11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...       $9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...       $9.25

In [237]:
# Get only the items that containe Chips 
chips = ds[ds.item_name.str.contains('chips', case=False)]

In [239]:
# Get the total quantity ordered for each thing
chips_cnt = chips.groupby('item_name').quantity.sum()

item_name
Chips                                    230
Chips and Fresh Tomato Salsa             130
Chips and Guacamole                      506
Chips and Mild Fresh Tomato Salsa          1
Chips and Roasted Chili Corn Salsa        23
Chips and Roasted Chili-Corn Salsa        18
Chips and Tomatillo Green Chili Salsa     45
Chips and Tomatillo Red Chili Salsa       50
Chips and Tomatillo-Green Chili Salsa     33
Chips and Tomatillo-Red Chili Salsa       25
Side of Chips                            110
Name: quantity, dtype: int64

In [236]:
# Let's add chips_cnt to the dict
chipsdict = {}
k=0
for i in chips_cnt.index:
    chipsdict[i] = chips_cnt[k]
    k+=1
chipsdict

{'Chips': 230,
 'Chips and Fresh Tomato Salsa': 130,
 'Chips and Guacamole': 506,
 'Chips and Mild Fresh Tomato Salsa': 1,
 'Chips and Roasted Chili Corn Salsa': 23,
 'Chips and Roasted Chili-Corn Salsa': 18,
 'Chips and Tomatillo Green Chili Salsa': 45,
 'Chips and Tomatillo Red Chili Salsa': 50,
 'Chips and Tomatillo-Green Chili Salsa': 33,
 'Chips and Tomatillo-Red Chili Salsa': 25,
 'Side of Chips': 110}

---

## Bonus: Craft a problem statement about this data that interests you, and then answer it!
